### DISCUSSION TITLE
Express.js app returns 404 error despite correct route configuration- Vercel deployment issue

### DISCUSSION THREAD

**Discussion by piotrjass** [vercel/community#5834] [Help] [answered]
### Summary

I have an app.js file in root directory

```
const express = require('express');
// Routes
const homeRouter = require('./routes/homeRoutes');
```
...
```
const app = express();
routers.forEach(router => {
  app.use('/api/v1', router);
});
module.exports = app;
module.exports.handler = serverless;
```
i deleted a lot of code but the idea is that it works.

homeRoutes.js
```

const express = require('express');

const router = express.Router();

router.get('/home', (req, res) => {
  res.send('Hello from home!');
});

module.exports = router;
```
vercel.json

```
{
  "routes": [
    {
      "src": "/",
      "dest": "/"
    },
    { "src": "/home", "dest": "/" }
  ]
}
```
package.json

```
 "name": "server",
  "version": "1.0.0",
  "engines": {
    "node": "18.x"
  },
```
And i cant deploy express app on Vercel. I tried following their docs, and different combination of Vercel.json but all the time i get:
```

404: NOT_FOUND
Code: NOT_FOUND
```
I can access locally `http://localhost:3000/api/v1/home` but the

`vercel-url/api/v1/home `or any comibation like url/home...

doesnt work and i get the the same error all the time. Different Vercel.json files that doesnt work:

```
{
  "routes": [{ "src": "/api/v1/home", "dest": "api/home.js" }]
}
```
or

```
{
  "rewrites": [{ "source": "/api/(.*)", "destination": "/api" }]
}
```

During deployment everything seems okey but then i cant access anything.
Do you know what can make this error or how to fix it?

### Steps to Reproduce

Depoly and go to ur;/api/v1/home

thats the file structure

```
d-----        13.02.2024     23:22                .parcel-cache
d-----        13.02.2024     13:00                controllers
d-----        13.02.2024     22:58                functions
d-----        13.02.2024     13:00                models
d-----        13.02.2024     23:35                node_modules
d-----        14.02.2024     23:30                routes
d-----        13.02.2024     13:00                utils
-a----        14.02.2024     23:03            354 .gitignore
-a----        21.09.2023     14:32             29 .prettierrc
-a----        14.02.2024     23:32           2699 app.js
-a----        24.09.2023     17:00            296 config.env
-a----        13.02.2024     23:35         947495 package-lock.json
-a----        15.02.2024     11:46           1428 package.json
-a----        01.10.2023     23:08            740 server.js
-a----        15.02.2024     00:25             70 vercel.json
```

server.js

```
// eslint-disable-next-line no-unused-vars
const dotenv = require('dotenv').config({ path: 'config.env' });
const http = require('http');
const mongoose = require('mongoose');
const app = require('./app');
// conne
const DB = process.env.DATABASE.replace(
  '<password>',
  process.env.DATABASE_PASSWORD
);

mongoose
  .connect(DB, {
    dbName: 'JSQuiz',
    useNewUrlParser: true,
    useCreateIndex: true,
    useFindAndModify: false,
    useUnifiedTopology: true
  })
  .then(() => console.log('DB connection successful!'));

const port = process.env.PORT;
// eslint-disable-next-line no-unused-vars

// eslint-disable-next-line no-unused-vars
const server = http.createServer(app);

server.listen(port, () => {
  console.log(`App running on port ${port}...`);
});
// const server = app.listen(port, () => {
//   console.log(`App running on port ${port}...`);
// });

```

i know the vercel docs says i shoud have root/api/index.js but i tried and it didnt work.

**Comment by amyegan** [comment, marked as answer]
Hey @piotrjass. Try moving your Express.js code into an `/api` directory at the root of the project. You can learn more about why this should work and how to deploy Express apps with Vercel in this guide: https://vercel.com/guides/using-express-with-vercel

**Reply by dheerajaggarwal2** [reply]
Hey @amyegan I am also facing the same issue. I am using the below structure. Not sure what I can do more here. I have tried all the ways.
This is the repo link:
https://github.com/dheerajaggarwal2/testing-bansal-backend-service

<img width="1437" alt="Screenshot 2024-02-26 at 4 14 47 PM" src="https://github.com/vercel/community/assets/20679277/a002de81-3f57-4fc3-9461-03e8a8615ff9">

**Reply by dheerajaggarwal2** [reply]
@piotrjass Were you able to resolve the issue?

**Reply by amyegan** [reply]
You shouldn't need `builds` and `routes` in the configuration if you have `rewrites` and the code is in `/api`. That could be complicating things.

I recently put together a minimal Express app based on the [guide](https://vercel.com/guides/using-express-with-vercel) in case it helps anyone to compare with an example app.
- Repo: https://github.com/amyegan/express-vercel
- Deployed example: https://express-vercel-woad.vercel.app/

**Reply by dheerajaggarwal2** [reply]
Thanks @amyegan I was able to resolve the issue. There was an issue with the code structure. Thanks for you input.

**Reply by n-zngr** [reply]
Hi, I'm facing the same issue. What did you do in order for it to work specifically? Cheers!

**Reply by amyegan** [reply]
I came up with a few ideas based on common problems I've seen.
- Your Express server code should be in `/api/index.js`
- A `vercel.json` file should be used to point all traffic to the `/api` directory
    ```json
    {
      "rewrites": [
        { "source": "/(.*)", "destination": "/api" }
      ]
    }
    ```
- The `vercel.json` file should be in the root directory, not a sub-folder within the project repo

**Reply by n-zngr** [reply]
Also helps when I figure out that I'm trying to run a server on Vercel, while Vercel simply doesn't support that on the free plan. Cheers. 
Maybe a bit more context on what I was trying to host: 
https://github.com/DeltaGamingCH/FINANCEZ
A platform where users can create new income/expense datasets, the server saves those through MongoDB and displays them accordingly.

**Reply by Rajukumarsaw** [reply]
Thanks @amyegan , finally my issue is resolved

**Comment by MuhammadIbtasam123** [comment]
Hey @amyegan I am also facing the same issue. I am using the below structure. I have tried all the above ways as well as the repo example you provide.
This is the repo link:
https://github.com/MuhammadIbtasam123/Ibtasam-portfolio
![Screenshot 2024-03-17 002033](https://github.com/vercel/community/assets/119099331/c8974331-640e-46b8-b9b5-c23b39f3bfef)

**Reply by MuhammadIbtasam123** [reply]
The project successfully deployed now. Some changes that i have made , if someone doing those read this.
in package.json if you are using nodemon on local server don't use it instead run you file uisng node <filename>, modify you package.json accordingly,

{
....
  "main": "api/index.js",

  "scripts": {
....
    "start": "node api/index.js"
  },
  "author": "",
  "license": "ISC",
  "keywords": [],
  "description": "",
  "dependencies": {
.....
  }
}
 index.js is my main entring point.
 
Secondly, @amyegan mention in her repo express-vercel guide, make sure make api folder and name file as index.js. 
Reference guide : https://github.com/amyegan/express-vercel
Make these changes to vercel.json
{
  "rewrites": [
    { "source": "/(.*)", "destination": "/api" }
  ]
}

This way it works for me. Thanks

**Comment by Rajukumarsaw** [comment]
why after refreshing some pages, it starts showing error message. 
![Screenshot_17-3-2024_184220_wowch vercel app](https://github.com/vercel/community/assets/90413057/18258e84-16bb-418d-80a7-a980617a911b)
![Screenshot_17-3-2024_184317_wowch vercel app](https://github.com/vercel/community/assets/90413057/d4bae50b-77aa-4843-a19e-a0b1a1e5fce2)
this two pictures are of same url
